pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,308 @@
+	for statusCollectorName, clauses := range statusCollectorNameToClauses {
+		if _, ok := c.statusCollectorNameToData[statusCollectorName]; !ok {
+			c.statusCollectorNameToData[statusCollectorName] = &statusCollectorData{
+				statusCollectorClauses: clauses,
+				workStatusRefToData:    make(map[util.ObjectIdentifier]workStatusData),
+			}
+
+			workStatusAdded = true
+		}
+	}
+
+	return workStatusRemoved, workStatusAdded
+}
+
+// getIdentifier returns the identifier of the combinedstatus resolution.
+// TODO: implement
+func (c *combinedStatusResolution) getIdentifier() util.ObjectIdentifier {
+	return util.ObjectIdentifier{}
+}
+
+// updateStatusCollector updates the status collector data in the
+// combinedstatus resolution. If the status collector is not relevant to the
+// latter, the function returns false. The function returns true if the status
+// collector data is updated. The given clauses are assumed to be valid and
+// immutable.
+func (c *combinedStatusResolution) updateStatusCollector(statusCollectorName string,
+	clauses *statusCollectorClauses) bool {
+	c.Lock()
+	defer c.Unlock()
+
+	scData, ok := c.statusCollectorNameToData[statusCollectorName]
+	if !ok {
+		return false // statusCollector is irrelevant to this combinedstatus resolution
+	}
+
+	if scData.equals(clauses) {
+		return false // statusCollector data is already up-to-date
+	}
+
+	// status collector clauses need to be updated, therefore update fields
+	// and invalidate all cached workstatus evaluations by resetting the map
+	c.statusCollectorNameToData[statusCollectorName] = &statusCollectorData{
+		statusCollectorClauses: clauses,
+		workStatusRefToData:    make(map[util.ObjectIdentifier]workStatusData),
+	}
+
+	return true
+}
+
+// evaluateWorkStatus evaluates the workstatus per relevant statuscollector.
+// The function returns true if any statuscollector data is updated.
+func (c *combinedStatusResolution) evaluateWorkStatus(ctx context.Context, celEvaluator *celEvaluator,
+	workStatusIdentifier util.ObjectIdentifier, workStatusContent *runtime.RawExtension) (bool, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	logger := klog.FromContext(ctx)
+
+	updated := false
+	for statusCollectorName := range c.statusCollectorNameToData {
+		scData, ok := c.statusCollectorNameToData[statusCollectorName]
+		if !ok {
+			continue
+		}
+
+		changed, err := c.evaluateWorkStatusAgainstStatusCollector(celEvaluator, workStatusIdentifier,
+			workStatusContent, scData)
+		if err != nil {
+			logger.Error(err, "failed to evaluate workstatus against statuscollector",
+				"statusCollectorName", statusCollectorName)
+			return false, nil
+		}
+
+		updated = updated || changed
+	}
+
+	return updated, nil
+}
+
+// evaluateWorkStatusAgainstStatusCollector evaluates the workstatus against
+// the statuscollector clauses and caches the evaluations. The function returns
+// true if an evaluation is updated.
+// If any evaluation fails, the function returns an error.
```

Comment:
The API design is a little more structured.

https://github.com/kubestellar/kubestellar/blob/v0.23.0-alpha.3/api/control/v1alpha1/types.go#L419-L420

https://github.com/kubestellar/kubestellar/blob/v0.23.0-alpha.3/api/control/v1alpha1/types.go#L353 . I will admit that I was assuming when I wrote that that an SQL-like treatment of "undefined" values and that it would be conflated with "null". I do not think that Eval errors should abort processing; I think that they should have defined effects on the fields of a StatusCollectorSpec. As I tried to say in this case, an undefined value (AKA Eval error) should be treated as failing to pass the filter.

Eval error in GroupBy should map to "the undefined value".
Same for Select.
For aggregations, Eval error in the NamedAggregator.Subject should exclude that row (WorkStatus) from the aggregation's input.

Reply:
I was under the impression that we'd want to abort the whole statement if part of it is bad, similarly to how most SQL dialects behave if im not mistaken.

Reply:
It is a design question worth discussing.
I was thinking that in a large dynamic system, there will be the occasional thing that is unexpected. I did not want that to ruin the aggregation of stuff with the expected shape. I also worried about the volume of complaints if every exception is reported.

Reply:
I think it's sensible to log these errors and not propagate them up the stack for now. Just to be clear, the failure of evaluating one workstatus against a single statuscollector should not affect other statuscollectors in the combinedstatus. 

It is highly likely that if one workstatus fails a statuscollector clause then all other associated workstatuses would as well. In the cases where this does not hold, aggregated status would still be calculated using the rest. I believe it's simpler to abort in this code iteration - what do you think?

Reply:
Fixed

Reply:
I think that the user should get some state-based reporting of evaluation errors.

Here is a proposal. The first question to answer is: how to the errors appear in the API? Let us divide the evaluation errors into two categories: (1) those that arise while creating the materialized views, and (2) those that arise while reducing those to the NamedStatusCombination structs. Note that these categories of error arise at different times: (1) when updating a materialized view, (2) when reducing such a view by aggregation (creating/updating a CombindStatus object). For category 1, perhaps we could extend NamedStatusCombination with the ability to report (a) one error of this category (including a display of the error and the name of the WEC that it comes from --- and insist on taking the error from the lexicographically first WEC with such an error, for the sake of determinism) and (b) a count of the number of rows in the materialized view that are missing due to evaluation errors. For category 2, perhaps we could also extend NamedStatusCombination with the ability to report one error per column, reporting the problem in terms of values without attempting to identify which row of the result would exist if the error had not arisen.

For category 1 errors, the implementation would extend workStatusData with the ability to hold one `error` of category 1.  The one that gets recorded is the first one that arises when trying to do the evaluations for that workStatusData. For category 2 errors, the data structure does not need to be extended.

Reply:
Will be addressed in followup PR - logged in issue #2141.